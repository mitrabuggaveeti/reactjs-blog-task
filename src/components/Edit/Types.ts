export interface IState{
    details : {
        id : string,
        userId : string,
        title : string,
        body : string
    }
} 

export interface ParamTypes {
    id: string;
    userId : string;
}
