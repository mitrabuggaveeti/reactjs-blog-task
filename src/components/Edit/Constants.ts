export const SUBMIT_BUTTON_LABEL = 'Submit'
export const INPUT_FIELD_LABEL = 'Title'
export const TEXT_FILED_LABEL = 'Body'
export const EDIT_MESSAGE = 'Edit Blog for UserID'
