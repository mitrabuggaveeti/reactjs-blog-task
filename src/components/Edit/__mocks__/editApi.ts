import { IState } from '../Types';


export const editPost = async (details : IState['details']) =>{
    return await new Promise(resolve =>{
        resolve({ok:true})
    })
 }
 