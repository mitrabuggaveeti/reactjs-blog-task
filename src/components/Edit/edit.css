
.editForm {
    display: flex;
    flex-direction: column;
    width: 30rem;
    margin: 5rem auto
}

.editForm-input {
    padding: 0.5rem;
    font-size: 1rem;
    margin: 0.3rem 0rem
}

.editForm-btn {
    padding: 0.5rem;
    cursor: pointer;
    background-color: #0b5468;
    font-weight: 700;
    color: white;
    border: none
}
