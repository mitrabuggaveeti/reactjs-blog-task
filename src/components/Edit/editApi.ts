import { IState } from './Types';

export const editPost = async (details : IState['details']) =>{

  try {
      const response = await fetch('https://jsonplaceholder.typicode.com/posts/1', {
        method: 'PUT',
        body: JSON.stringify(details),
        headers: {
          'Content-type': 'application/json; charset=UTF-8',
        },
      })
    return await response.json();
  } catch (error) {
     return {error}
  }

}

