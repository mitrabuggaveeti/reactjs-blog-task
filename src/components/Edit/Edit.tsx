import React, { useState } from 'react'
import './edit.css';
import {useParams, useHistory} from 'react-router-dom';
import {editPost} from './editApi';
import {IState, ParamTypes} from './Types';

const Edit : React.FC = () => {

    const params = useParams<ParamTypes>();
    let history = useHistory();
    

    const [input, setInput] = useState<IState['details']>({
        ...params,
        title: "",
        body: "",
    }) 
 
    const handleChange = (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
        setInput({
            ...input,
            [e.target.name]: e.target.value
        })
    }

    const handleClick = async() =>{
        const data = await editPost(input);
        history.push('/feed');
    }

    return (
        <div className="editForm">
            <strong>Edit Blog for UserID : {params.userId} & PostID : {params.id}</strong>
            <input 
                type="text"
                onChange={handleChange}
                className="editForm-input"
                name="title"
                value={input.title}
                placeholder="Title"
            />
            <textarea
                onChange={handleChange}
                className="editForm-input"
                name="body"
                value={input.body}
                placeholder="Body"
            />

            <button
                onClick={handleClick}
                className="editForm-btn"
                data-testid='submit'
            >
                Submit
            </button>
     </div>
    )
}

export default Edit
