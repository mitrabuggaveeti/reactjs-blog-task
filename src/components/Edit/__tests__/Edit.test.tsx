import { cleanup, fireEvent, render,waitFor} from "@testing-library/react";
import Edit from "../Edit";
import * as services from '../editApi';

jest.mock('../editApi');

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({
      id: '1',
      userId: '2',
    }),
}));

describe('Edit component should', () => {
    afterEach(cleanup);

    it('render',async () => {
                
        const { asFragment } = render(<Edit/>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('Make edit api reqquest on submit', async ()=>{

        const mockAddListener = jest.spyOn(services, 'editPost');
        const { getByTestId} = render(<Edit/>);

        fireEvent.click(getByTestId('submit'));
        expect(mockAddListener).toHaveBeenCalled();
    })

})
