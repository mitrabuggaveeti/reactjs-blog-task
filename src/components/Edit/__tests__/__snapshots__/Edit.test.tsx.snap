// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit component should render 1`] = `
<DocumentFragment>
  <div
    class="editForm"
  >
    <strong>
      Edit Blog for UserID : 2 & PostID : 1
    </strong>
    <input
      class="editForm-input"
      name="title"
      placeholder="Title"
      type="text"
      value=""
    />
    <textarea
      class="editForm-input"
      name="body"
      placeholder="Body"
    />
    <button
      class="editForm-btn"
      data-testid="submit"
    >
      Submit
    </button>
  </div>
</DocumentFragment>
`;
